core/executor.py

```diff
@@ -595,41 +595,62 @@ def RunSubshell(self, node):
         return p.RunProcess(self.waiter, trace.ForkWait)
 
     def CaptureStdout(self, node):
-        # type: (command_t) -> Tuple[int, str]
+        # type: (command_t) -> Tuple[int, str, str]
 
         p = self._MakeProcess(node, self.exec_opts.inherit_errexit(),
                               self.exec_opts.errtrace())
         # Shell quirk: Command subs remain part of the shell's process group, so we
         # don't use p.AddStateChange(process.SetPgid(...))
 
         r, w = posix.pipe()
+        r2, w2 = posix.pipe()
         p.AddStateChange(process.StdoutToPipe(r, w))
+        p.AddStateChange(process.StderrToPipe(r2, w2))
 
         p.StartProcess(trace.CommandSub)
         #log('Command sub started %d', pid)
 
         chunks = []  # type: List[str]
+        chunks2 = []  # type: List[str]
         posix.close(w)  # not going to write
+        posix.close(w2)  # not going to write
+        eofcount = 0
+        finished = False
         while True:
-            n, err_num = pyos.Read(r, 4096, chunks)
+            fds = pyos.WaitForInputs([r, r2])
 
-            if n < 0:
-                if err_num == EINTR:
-                    pass  # retry
-                else:
-                    # Like the top level IOError handler
-                    e_die_status(
-                        2,
-                        'Oils I/O error (read): %s' % posix.strerror(err_num))
+            # zero outputs mean something went wrong
+            if len(fds) == 0:
+                break
 
-            elif n == 0:  # EOF
+            for fd in fds:
+                if fd == r:
```

Comment:
what is this condition for?

Reply:
Made it more clear now. It checks if it's the `stdout_fd` and writes to the `stdout_chunks`, otherwise to the `stderr_chunks`.